database/pgsql/layer.go

```diff
@@ -238,28 +250,51 @@ func (pgSQL *pgSQL) InsertLayer(layer database.Layer) error {
 
 	// Get parent ID.
 	var parentID zero.Int
+	mapNamespaceIDs := make(map[string]int)
 	if layer.Parent != nil {
 		if layer.Parent.ID == 0 {
 			log.Warning("Parent is expected to be retrieved from database when inserting a layer.")
 			return cerrors.NewBadRequestError("Parent is expected to be retrieved from database when inserting a layer.")
 		}
 
 		parentID = zero.IntFrom(int64(layer.Parent.ID))
-	}
 
-	// Find or insert namespace if provided.
-	var namespaceID zero.Int
-	if layer.Namespace != nil {
-		n, err := pgSQL.insertNamespace(*layer.Namespace)
+		// Find its parent's namespaces
+		rows, err := pgSQL.Query(searchLayerNamespace, parentID)
 		if err != nil {
 			return err
 		}
-		namespaceID = zero.IntFrom(int64(n))
-	} else if layer.Namespace == nil && layer.Parent != nil {
-		// Import the Namespace from the parent if it has one and this layer doesn't specify one.
-		if layer.Parent.Namespace != nil {
-			namespaceID = zero.IntFrom(int64(layer.Parent.Namespace.ID))
+		defer rows.Close()
+
+		for rows.Next() {
+			var pn database.Namespace
+
+			err = rows.Scan(&pn.ID, &pn.Name)
+			if err != nil {
+				return err
+			}
+
+			name := strings.Split(pn.Name, ":")
+			// Layer will inherit its parent's namespace
+			mapNamespaceIDs[name[0]] = pn.ID
+		}
+		if err = rows.Err(); err != nil {
+			return err
+		}
+	}
+
+	// Insert Namespaces
+	for _, n := range layer.Namespaces {
+		id, err := pgSQL.insertNamespace(n)
+		if err != nil {
+			return err
 		}
+
+		// Layer's namespaces has high priority than its parent.
+		// Once a layer has a 'same' namespace (the content before ':' is the same)
+		// with its parent, it will only keep its namespace.
+		name := strings.Split(n.Name, ":")
+		mapNamespaceIDs[name[0]] = id
```

Comment:
I understand what you are doing here and this is absolutely right. I think that instead of doing this split around `:`, we should probably extend `Namespace`. This is quite a big design decision and we need to discuss more with @jzelinskie about it.